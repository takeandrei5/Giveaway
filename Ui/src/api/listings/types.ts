import { PaginatedResult } from '../../utils/types';

export type ItemData = { id: string; image: string; title: string; createdAt: Date };

export type CreateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export type UpdateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export interface FetchListingsResponse {
	listings: PaginatedResult<ItemData>;
}

export type ListingInformation = {
	id: string;
	title: string;
	description: string;
	category: number;
	createdAt: Date;
	images: string[];
};

export type OwnerInformation = {
	email: string;
	name: string;
	image: string;
};

export type FetchListingDetailsResponse = { listingInfo: ListingInformation } & { ownerInfo: OwnerInformation };
