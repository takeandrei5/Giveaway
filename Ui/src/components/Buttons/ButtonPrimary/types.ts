import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = Omit<ButtonProps, 'color'>;
