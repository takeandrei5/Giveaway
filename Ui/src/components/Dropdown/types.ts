export type DropdownProps = {
	id: string;
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
	name: string;
	value: string | number | readonly string[];
	isInvalid?: false;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
