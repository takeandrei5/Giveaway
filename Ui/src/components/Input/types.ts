export type InputProps = {
	id: string;
	name: string;
	placeholder: string;
	value?: string;
	onBlur?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onChange?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyUp?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	disabled?: boolean;
	multiline?: boolean;
	isInvalid?: boolean;
	height?: string | number;
	label?: string;
	leftIcon?: JSX.Element;
	rightIcon?: JSX.Element;
	rows?: number;
	width?: string | number;
};
