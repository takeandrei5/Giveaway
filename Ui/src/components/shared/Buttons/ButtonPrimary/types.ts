import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = ButtonProps;
