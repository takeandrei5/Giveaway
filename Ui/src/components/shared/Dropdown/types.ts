export type DropdownProps = {
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
