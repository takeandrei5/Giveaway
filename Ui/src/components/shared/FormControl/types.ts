export type FormControlProps = {
	children: React.ReactElement;
	label: string;
	id: string;
	name: string;
};

export type useCheckFormIsInvalidProps = {
	isInvalid: boolean;
};
