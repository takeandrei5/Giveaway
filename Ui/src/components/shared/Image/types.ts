import { ImageProps as NextImageProps } from 'next/image';

export type ImageProps = NextImageProps & {
	height: string;
	width: string;
	backgroundColor?: string;
	borderRadius?: string;
	draggable?: boolean;
};
