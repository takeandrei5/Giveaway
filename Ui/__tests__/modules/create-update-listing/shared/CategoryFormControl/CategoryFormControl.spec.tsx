import { CategoryFormControl } from '@modules/create-update-listing/shared/CategoryFormControl';
import { CategoryFormControlProps } from '@modules/create-update-listing/shared/CategoryFormControl/types';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useField } from 'formik';

jest.mock('formik', () => ({
	...jest.requireActual('formik'),
	useField: jest.fn(),
}));

describe('CategoryFormControl', () => {
	it('should match snapshot', () => {
		// Arrange
		const props: CategoryFormControlProps = {
			name: 'test-name',
		};

		(useField as unknown as jest.Mock).mockImplementation(() => [
			{
				value: 'test-name',
			},
			{
				value: 'test-name',
				error: '',
				touched: false,
				initialValue: 'test-name',
				initialTouched: false,
				initialError: '',
			},
		]);

		// Act
		const { container } = render(<CategoryFormControl {...props} />);

		// Assert
		expect(container).toMatchSnapshot();
	});

	it('should execute onChangeHandler correctly', async () => {
		// Arrange
		const props: CategoryFormControlProps = {
			name: 'test-name',
		};

		const setValueMock = jest.fn<{ value: boolean; shouldValidate?: boolean }, any>();
		const setTouchedMock = jest.fn<{ value: boolean; shouldValidate?: boolean }, any>();

		(useField as unknown as jest.Mock).mockImplementation(() => [
			{
				value: 'test-name',
			},
			{
				value: 'test-name',
				error: '',
				touched: false,
				initialValue: 'test-name',
				initialTouched: false,
				initialError: '',
			},
			{
				setValue: setValueMock,
				setTouched: setTouchedMock,
			},
		]);

		// Act
		const { getByTestId } = render(<CategoryFormControl {...props} />);

		await userEvent.selectOptions(getByTestId('dropdown'), getByTestId('dropdown').querySelector('option')!.value);

		// Assert
		expect(setValueMock).toHaveBeenCalledTimes(1);
		expect(setTouchedMock).toHaveBeenCalledTimes(1);
	});
});
