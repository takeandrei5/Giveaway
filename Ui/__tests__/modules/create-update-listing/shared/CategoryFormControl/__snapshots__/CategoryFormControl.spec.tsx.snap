// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryFormControl should match snapshot 1`] = `
<div>
  <div
    class="chakra-form-control css-8atqhb"
    role="group"
  >
    <label
      class="chakra-form__label css-xgxc5v"
      for="category-dropdown"
      id="field-1-label"
    >
      Category:
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
    <div
      class="chakra-select__wrapper css-yumodq"
    >
      <select
        aria-required="true"
        class="chakra-select css-1inymzc"
        data-testid="dropdown"
        id="category-dropdown"
        name="test-name"
        required=""
      >
        <option
          value="1"
        >
          Clothes
        </option>
        <option
          value="2"
        >
          Toys
        </option>
        <option
          value="3"
        >
          Books
        </option>
        <option
          value="4"
        >
          Electronics
        </option>
      </select>
      <div
        class="chakra-select__icon-wrapper css-6fsspc"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
